utils/dom.js

```diff
@@ -140,11 +136,47 @@ export function isVerticalEdge( container, isReverse, collapseRanges = false ) {
 	return true;
 }
 
+/**
+ * Get the rectangle of a given Range.
+ *
+ * @param {Range} range The range.
+ *
+ * @return {DOMRect} The rectangle.
+ */
+export function getRectangleFromRange( range ) {
+	// For uncollapsed ranges, get the rectangle that bounds the contents of the
+	// range; this a rectangle enclosing the union of the bounding rectangles
+	// for all the elements in the range.
+	if ( ! range.collapsed ) {
+		return range.getBoundingClientRect();
+	}
+
+	// If the collapsed range starts (and therefore ends) at an element node,
+	// `getClientRects` will return undefined. To fix this we can get the
+	// bounding rectangle of the element node to create a DOMRect based on that.
+	if ( range.startContainer.nodeType === ELEMENT_NODE ) {
+		const { x, y, height } = range.startContainer.getBoundingClientRect();
+
+		// Create a new DOMRect with zero width.
+		return new DOMRect( x, y, 0, height );
+	}
+
+	// For normal collapsed ranges (exception above), the bounding rectangle of
+	// the range may be inaccurate in some browsers. There will only be one
+	// rectangle since it is a collapsed range, so it is safe to pass this as
+	// the union of them. This works consistently in all browsers.
+	return first( range.getClientRects() );
```

Comment:
In testing #5769, I encountered a case where this revised implementation would return `undefined`, but the previous implementation would provide a valid `DOMRect`. Should be reproducible by merely clicking into one of the first paragraphs of the demo content.

Reply:
Thanks for testing, I'll have a look at this right now.

Reply:
@aduth I've tried a few browsers but cannot reproduce. I always get a list of one item in Chrome, Safari and Firefox. Which browser are you testing and what are the steps to reproduce the problem?

To ensure a `DOMRect` is always returned, we could pass an empty one, so I'm wondering if the previous implementation returned an accurate one...